### PR TITLE
feat: indicating generators are being installed in background

### PR DESCRIPTION
**Pull request by alex-gilin**
image

image

### CODE REVIEW

File: backend/src/panels/YeomanUIPanel.ts [hunk 1]
```diff
@@ -31,15 +31,23 @@ export class YeomanUIPanel extends AbstractWebviewPanel {
...
 		const yeomanUi = _.get(this, "yeomanui");
 		if (yeomanUi) {
-			yeomanUi._notifyGeneratorsChange(YeomanUIPanel.getGensMeta());
+			if (_.isNil(this.installGens)) {
+				yeomanUi._notifyGeneratorsChange(YeomanUIPanel.getGensMeta());
+			} else {
+				yeomanUi._notifyGeneratorsInstall(this.installGens);
+				if (_.isEmpty(this.installGens)) {
+					this.installGens =  undefined;
+				}
```
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/src/panels/YeomanUIPanel.ts:49]: if this.installGens === [] then it means that AppStudio completed its generators installations.

[the remark this answers is not in the document]
Comment (i300543) [reply backend/src/panels/YeomanUIPanel.ts:49]: why do you need to set it to undefined in this case ?

[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/src/panels/YeomanUIPanel.ts:49]: so next time generators prompt appears you won't get the 'complete installing' message.

File: backend/src/panels/YeomanUIPanel.ts [hunk 2]
```diff
@@ -78,6 +86,7 @@ export class YeomanUIPanel extends AbstractWebviewPanel {
 
 	private yeomanui: YeomanUI;
 	private messages: any;
+	private installGens: any;
```
Comment (alex-gilin) [inline backend/src/panels/YeomanUIPanel.ts:90]: I see before you store an array in this variable

File: backend/src/yeomanui.ts [hunk 3]
```diff
@@ -116,6 +116,24 @@ export class YeomanUI {
...
 
+	public async _notifyGeneratorsInstall(args: any[], force: boolean) {
+		if (!_.isNil(args)) {
+			const isGeneratorsPrompt: boolean = await this.rpc.invoke("isGeneratorsPrompt");
+			if (isGeneratorsPrompt || force) {
+				const message = this.getGeneratorsInstallingMessage(args);
+				this.youiEvents.getAppWizard().showInformation(message, MessageType.prompt);
+			}
+		}
+		this.uiOptions.installGens = (_.isObject(args) && _.isEmpty(args)) ? undefined : args;
```
Comment (alex-gilin) [inline backend/src/yeomanui.ts:127]: can you explain, what is expected result here ? _.isObject() is true on arrays, functions, objects, regexes, new Number(0), and new String('')

File: backend/src/yeomanui.ts [hunk 4]
```diff
@@ -116,6 +116,24 @@ export class YeomanUI {
...
+	}
+
+	private getGeneratorsInstallingMessage(args: any[]): string {
+		if (_.isEmpty(args)) {
+			return "We have finished installing all generators, enjoy your work!";
```
Comment (alex-gilin) [inline backend/src/yeomanui.ts:132]: Did Paola check the text of the message?

File: backend/src/yeomanui.ts [hunk 5]
```diff
@@ -116,6 +116,24 @@ export class YeomanUI {
...
+		}
+		return "There are some generators that are being installed.";
```
Comment (alex-gilin) [inline backend/src/yeomanui.ts:134]: same as above

File: backend/tests/yeomanui.spec.ts [hunk 6]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
 		});
 	});
 
+	describe("_notifyGeneratorsInstall", () => {
+		it("args = undefined", async () => {
+			rpcMock.expects("invoke").withExactArgs("isGeneratorsPrompt").never();
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:271]: seem like redundant

File: backend/tests/yeomanui.spec.ts [hunk 7]
```diff
@@ -221,6 +239,13 @@ describe('yeomanui unit test', () => {
 		});
 	});
 
+	describe("getState", () => {
+		it("getState", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:243]: please change the unit tests description to be more clear - expected result, flow description and may be parameters. 
very difficult to understand what test meaning and what its purpose.

File: backend/tests/yeomanui.spec.ts [hunk 8]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
 		});
 	});
 
+	describe("_notifyGeneratorsInstall", () => {
+		it("args = undefined", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:270]: please provide the clear unit test description

File: backend/tests/yeomanui.spec.ts [hunk 9]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
...
+			rpcMock.expects("invoke").withExactArgs("isGeneratorsPrompt").never();
+			await yeomanUi._notifyGeneratorsInstall(undefined, false);
+			expect(yeomanUi["uiOptions"].installGens).to.be.undefined;
+		});
+		it("args = [], isGeneratorsPrompt = true , force = false", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:275]: please provide the clear unit test description

File: backend/tests/yeomanui.spec.ts [hunk 10]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
...
+			const args: any = [];
+			rpcMock.expects("invoke").withExactArgs("isGeneratorsPrompt").resolves(true);
+			youiEventsMock.expects("getAppWizard").returns(appWizard);
+			appWizardMock.expects("showInformation").withExactArgs("We have finished installing all generators, enjoy your work!", MessageType.prompt);
+			await yeomanUi._notifyGeneratorsInstall(args, false);
+			expect(yeomanUi["uiOptions"].installGens).to.be.undefined;
+		});
+		it("isGeneratorsPrompt = true , force = false", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:283]: please provide the clear unit test description

File: backend/tests/yeomanui.spec.ts [hunk 11]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
...
+					versionRange: "^2.0.1"
+				}
+			];
+			rpcMock.expects("invoke").withExactArgs("isGeneratorsPrompt").resolves(true);
+			youiEventsMock.expects("getAppWizard").returns(appWizard);
+			appWizardMock.expects("showInformation").withExactArgs("There are some generators that are being installed.", MessageType.prompt);
+			await yeomanUi._notifyGeneratorsInstall(args, false);
+			expect(yeomanUi["uiOptions"].installGens).to.be.deep.equal(args);
+		});
+		it("isGeneratorsPrompt = false , force = true", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:304]: please provide the clear unit test description

File: backend/tests/yeomanui.spec.ts [hunk 12]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
...
+					versionRange: "^2.0.1"
+				}
+			];
+			rpcMock.expects("invoke").withExactArgs("isGeneratorsPrompt").resolves(false);
+			youiEventsMock.expects("getAppWizard").returns(appWizard);
+			appWizardMock.expects("showInformation").withExactArgs("There are some generators that are being installed.", MessageType.prompt);
+			await yeomanUi._notifyGeneratorsInstall(args, true);
+			expect(yeomanUi["uiOptions"].installGens).to.be.deep.equal(args);
+		});
+		it("isGeneratorsPrompt = false , force = false", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:325]: please provide the clear unit test description

File: backend/tests/yeomanui.spec.ts [hunk 13]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
...
+			rpcMock.expects("invoke").withExactArgs("isGeneratorsPrompt").resolves(false);
+			youiEventsMock.expects("getAppWizard").never();
+			appWizardMock.expects("showInformation").never();
+			await yeomanUi._notifyGeneratorsInstall(args, false);
+			expect(yeomanUi["uiOptions"].installGens).to.be.deep.equal(args);
+		});
+	});
+
+	describe("getGeneratorsInstallingMessage", async () => {
+		it("args = []", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:349]: please provide the clear unit test description

File: backend/tests/yeomanui.spec.ts [hunk 14]
```diff
@@ -241,6 +266,96 @@ describe('yeomanui unit test', () => {
...
+			const result = await yeomanUi["getGeneratorsInstallingMessage"]([]);
+			expect(result).to.be.deep.equal("We have finished installing all generators, enjoy your work!");
+		});
+		it("args = not empty", async () => {
```
Comment (alex-gilin) [inline backend/tests/yeomanui.spec.ts:353]: please provide the clear unit test description

File: frontend/tests/App.spec.js [hunk 15]
```diff
@@ -102,6 +102,37 @@ describe('App.vue', () => {
 		})
 	});
 
+	describe('isGeneratorsPrompt - method', () => {
+		it('promptIndex = 0, selectGeneratorPromptExists = true', () => {
```
Comment (alex-gilin) [inline frontend/tests/App.spec.js:106]: please provide the clear unit test description

File: frontend/tests/App.spec.js [hunk 16]
```diff
@@ -102,6 +102,37 @@ describe('App.vue', () => {
...
+			wrapper = initComponent(App, {}, true)
+			wrapper.vm.promptIndex = 0;
+			wrapper.vm.selectGeneratorPromptExists = () => true;
+			const result = wrapper.vm.isGeneratorsPrompt();
+			expect(result).toBeTruthy();
+		})
+		it('promptIndex = 0, selectGeneratorPromptExists = false', () => {
```
Comment (alex-gilin) [inline frontend/tests/App.spec.js:113]: please provide the clear unit test description

File: frontend/tests/App.spec.js [hunk 17]
```diff
@@ -102,6 +102,37 @@ describe('App.vue', () => {
...
+			wrapper = initComponent(App, {}, true)
+			wrapper.vm.promptIndex = 0;
+			wrapper.vm.selectGeneratorPromptExists = () => false;
+			const result = wrapper.vm.isGeneratorsPrompt();
+			expect(result).toBeFalsy();
+		})
+		it('promptIndex > 0, selectGeneratorPromptExists = true', () => {
```
Comment (alex-gilin) [inline frontend/tests/App.spec.js:120]: please provide the clear unit test description

File: frontend/tests/App.spec.js [hunk 18]
```diff
@@ -102,6 +102,37 @@ describe('App.vue', () => {
...
+			wrapper = initComponent(App, {}, true)
+			wrapper.vm.promptIndex = 1;
+			wrapper.vm.selectGeneratorPromptExists = () => true;
+			const result = wrapper.vm.isGeneratorsPrompt();
+			expect(result).toBeFalsy();
+		})
+		it('promptIndex > 0, selectGeneratorPromptExists = false', () => {
```
Comment (alex-gilin) [inline frontend/tests/App.spec.js:127]: please provide the clear unit test description

File: backend/src/yeomanui.ts (hunk 4 above)
Comment (alex-gilin) [inline backend/src/yeomanui.ts:132]: Is it worth to extract this message to messages file ?

File: backend/package.json [hunk 19]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "yeoman-ui",
-	"version": "1.1.56",
+	"version": "1.1.57",
```
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/package.json:3]: PATCH version is ok in this case.

File: backend/src/panels/YeomanUIPanel.ts [hunk 20]
```diff
@@ -31,15 +31,23 @@ export class YeomanUIPanel extends AbstractWebviewPanel {
...
+	public notifyGeneratorsChange(args?: any[]) {
+		this.installGens = !_.isObject(args) ? undefined : args;
 		const yeomanUi = _.get(this, "yeomanui");
 		if (yeomanUi) {
-			yeomanUi._notifyGeneratorsChange(YeomanUIPanel.getGensMeta());
+			if (_.isNil(this.installGens)) {
+				yeomanUi._notifyGeneratorsChange(YeomanUIPanel.getGensMeta());
+			} else {
+				yeomanUi._notifyGeneratorsInstall(this.installGens);
+				if (_.isEmpty(this.installGens)) {
```
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/src/panels/YeomanUIPanel.ts:47]: not redundant. _.isEmpty verify that it's empty obj --> []

File: backend/src/yeomanui.ts (hunk 4 above)
Comment (tomer-epstein) [reply backend/src/yeomanui.ts:132]: Yes. after Paola will review. + in a new PR I'll try notifying which gens are being installed.

File: backend/src/panels/YeomanUIPanel.ts [hunk 21]
```diff
@@ -31,15 +31,23 @@ export class YeomanUIPanel extends AbstractWebviewPanel {
...
 		this.output.show();
 	}
 
-	public notifyGeneratorsChange() {
+	public notifyGeneratorsChange(args?: any[]) {
+		this.installGens = !_.isObject(args) ? undefined : args;
 		const yeomanUi = _.get(this, "yeomanui");
 		if (yeomanUi) {
-			yeomanUi._notifyGeneratorsChange(YeomanUIPanel.getGensMeta());
+			if (_.isNil(this.installGens)) {
```
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/src/panels/YeomanUIPanel.ts]: if (!this.installGens) {

**Review by rimasirich — APPROVED** [review]

**Pull request merged**